File: Experiment 1/Experiment 2/merge_sort.cpp
/*
#include <iostream>
using namespace std;

// The function that merges two sorted halves
void merge(int arr[], int low, int mid, int high) {
    int n1 = mid - low + 1;
    int n2 = high - mid;

    // Create temporary arrays
    int L[n1], R[n2];

    // Copy data to temp arrays L[] and R[]
    for (int i = 0; i < n1; i++) L[i] = arr[low + i];
    for (int j = 0; j < n2; j++) R[j] = arr[mid + 1 + j];

    int i = 0, j = 0, k = low;

    // Merge the temp arrays back into arr[low..high]
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    // Copy any remaining elements of L[]
    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }

    // Copy any remaining elements of R[]
    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }
}

// The main Recursive function
void mergeSort(int arr[], int low, int high) {
    if (low < high) {
        int mid = low + (high - low) / 2;

        // Sort first and second halves
        mergeSort(arr, low, mid);
        mergeSort(arr, mid + 1, high);

        // Merge the sorted halves
        merge(arr, low, mid, high);
    }
}

int main() {
    int data[] = {38, 27, 43, 3, 9, 82, 10};
    int n = sizeof(data) / sizeof(data[0]);

    mergeSort(data, 0, n - 1);

    for (int i = 0; i < n; i++) {
        cout << data[i] << " ";
    }
    return 0;
}
*/

/*INSTRUCTIONS IF YOU WANT TO WRITE THE CODE YOURSELF.
1 Include the iostream library and use the std namespace
2 Create a void method 'merge' taking (int arr[], int low, int mid, int high)
3 Calculate sizes n1 (mid - low + 1) and n2 (high - mid)
4 Create temporary arrays L[n1] and R[n2]
5 Use for loops to copy data from the main array into L and R
6 Create pointers i=0, j=0, and k=low
7 Start a while loop that runs as long as i < n1 AND j < n2
8 Inside, if L[i] <= R[j], set arr[k] to L[i] and increment i
9 Else, set arr[k] to R[j] and increment j. Increment k after either choice.
10 Use two separate while loops to copy any remaining elements from L and R back to arr
11 Create a void method 'mergeSort' taking (int arr[], int low, int high)
12 Add a base case: if low is less than high, proceed
13 Calculate mid using low + (high - low) / 2
14 Recursively call mergeSort for the left half (low to mid)
15 Recursively call mergeSort for the right half (mid + 1 to high)
16 Call the merge function to combine the sorted results
17 In main, create an array, calculate its size, call mergeSort(data, 0, n-1), and print it
*/



#include <iostream>
using namespace std;

// 1. Function to merge two sorted sub-arrays
void merge(int arr[], int low, int mid, int high) {
    // 2. Define sizes of temporary arrays
    
    // 3. Create temp arrays L[] and R[]
    
    // 4. Copy data from main array to L and R
    
    // 5. Initialize pointers i, j, and k (k starts at low)
    
    // 6. While loop to compare elements and merge
    while (/* i is valid */ && /* j is valid */) {
        // If L[i] is smaller, put it in arr[k]
        // Else, put R[j] in arr[k]
    }
    
    // 7. Write while loops to catch any leftover elements in L or R
}

// 8. Recursive function to split the array
void mergeSort(int arr[], int low, int high) {
    if (low < high) {
        // 9. Calculate the mid point
        
        // 10. Recursive call for the left side
        
        // 11. Recursive call for the right side
        
        // 12. Merge the two sides together
    }
}

int main() {
    int data[] = {38, 27, 43, 3, 9, 82, 10};
    int n = sizeof(data) / sizeof(data[0]);

    // 13. Call mergeSort with the correct boundaries (low = 0, high = n-1)
    
    // 14. Print the sorted array using a for loop
    
    return 0;
}

File: Experiment 1/Experiment 2/merge_sort.py
"""
def merge_sort(arr):
    if len(arr) > 1:
        # Finding the mid of the array
        mid = len(arr) // 2

        # Dividing the array elements into 2 halves
        L = arr[:mid]
        R = arr[mid:]

        # Sorting the first half
        merge_sort(L)

        # Sorting the second half
        merge_sort(R)

        i = j = k = 0

        # Copy data to temp arrays L[] and R[]
        while i < len(L) and j < len(R):
            if L[i] <= R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1

        # Checking if any element was left
        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1

        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1

# Testing the code
data = [38, 27, 43, 3, 9, 82, 10]
merge_sort(data)
print(data)
"""

"""INSTRUCTIONS IF YOU WANT TO WRITE THE CODE YOURSELF.
1 Define a function named merge_sort that takes one argument: arr
2 Add a base case check: if the length of arr is greater than 1, proceed
3 Calculate the mid point using floor division (// 2)
4 Use Python slicing to create the left half (L) from start to mid
5 Use Python slicing to create the right half (R) from mid to end
6 Recursively call merge_sort on L
7 Recursively call merge_sort on R
8 Initialize three pointers: i, j, and k all to 0
9 Start a while loop that runs while i < len(L) and j < len(R)
10 Inside, compare L[i] and R[j]; place the smaller one into arr[k]
11 Increment the respective pointer (i or j) and the main pointer k
12 Write two separate while loops to copy any remaining elements from L and R
13 Outside the function, create a list named 'data' with unsorted numbers
14 Call merge_sort(data)
15 Print the final sorted list
"""

# 1. Define the merge_sort function
def merge_sort(arr):
    # 2. Check if the list has more than one element
    if len(arr) > 1:
        
        # 3. Calculate mid and slice the list into L and R halves
        
        # 4. Recursive calls for both halves
        
        # 5. Initialize i, j, k to zero
        i = j = k = 0
        
        # 6. The main comparison loop
        while i < len(L) and j < len(R):
            # If L[i] is smaller, arr[k] = L[i]
            # Else, arr[k] = R[j]
            pass
            
        # 7. Catch the remaining elements in L or R
        # (Hint: Use two separate while loops)

# 8. Test the implementation
data = [38, 27, 43, 3, 9, 82, 10]
# Call the function here
# Print the results